lib/Dialect/FIRRTL/Transforms/AddSeqMemPorts.cpp

```diff
@@ -403,27 +417,75 @@ void AddSeqMemPortsPass::runOnOperation() {
   }
   extraPortsAttr = ArrayAttr::get(context, extraPorts);
 
+  auto *effectiveDut = instanceInfo->getEffectiveDut();
+
   // If there are no user ports, don't do anything.
   if (!userPorts.empty()) {
     // Update ports statistic.
     numAddedPorts += userPorts.size();
 
-    // Visit the nodes in post-order.
-    for (auto *node : llvm::post_order(dutNode)) {
+    // Visit the modules in post-order in one of two ways:
+    //
+    //   1. If no design-under-test is specified, visit all modules.
+    //   2. If the design-under-test is specified, only visit modules under it.
+    //
+    // Additionally, skip any modules that are wholly instantiated under layers.
+    // If any memories are partially instantiated under a layer then error.
+    for (auto *node : llvm::post_order(instanceGraph)) {
       auto op = node->getModule();
+      // Skip the module if there is a DUT and no instances are under it.
+      if (instanceInfo->hasDut() && !instanceInfo->anyInstanceUnderDut(op))
+        continue;
+
+      // Skip anything wholly under a layer.
+      if (instanceInfo->allInstancesUnderLayer(op))
+        continue;
+
+      // Process the module or memory.  Error if the memory is partially under a
+      // layer.
+      //
+      // TODO: There are several ways to handle a memory being under a layer:
+      // duplication or tie-off.  However, it is unclear if either if these is
+      // intended or correct.  Additionally, this can be handled with pass order
+      // by preventing deduplication of memories that have this property in
+      // `LowerMemories`.
       if (auto moduleOp = dyn_cast<FModuleOp>(*op)) {
-        if (failed(processModule(moduleOp, /*isDUT=*/node == dutNode)))
+        if (failed(processModule(moduleOp)))
           return signalPassFailure();
       } else if (auto mem = dyn_cast<FMemModuleOp>(*op)) {
+        if (instanceInfo->anyInstanceUnderLayer(mem)) {
+          auto diag = op->emitOpError()
+                      << "cannot have ports added to it because it is "
+                         "instantiated under "
+                         "both the design-under-test and layer blocks";
+          for (auto *instNode : node->uses()) {
+            auto instanceOp = instNode->getInstance();
+            if (auto layerBlockOp =
+                    instanceOp->getParentOfType<LayerBlockOp>()) {
+              diag.attachNote(instanceOp.getLoc())
+                  << "this instance is under a layer block";
+              diag.attachNote(layerBlockOp.getLoc())
+                  << "the innermost layer block is here";
+              continue;
+            }
+            if (instanceInfo->anyInstanceUnderLayer(
+                    instNode->getParent()->getModule())) {
+              diag.attachNote(instanceOp.getLoc())
+                  << "this instance is inside a module that is instantiated "
+                     "under a layer block";
+            }
+          }
+          return signalPassFailure();
```

Comment:
I think this check belongs in processMemModule, which should be return a LogicalResult.

Reply:
Nice idea. This also makes this have a consistent signature with `processModule`. Changed in 2e5dd7fe8.